### ISSUE TITLE
[Feature Request] Ability to omit export when a value or type not used outside of the F# module/file

### ISSUE THREAD

**Issue by Zaid-Ajaj** [fable-compiler/Fable#2286]
### Description

This might be the final hurdle of getting React tooling (fast-refresh) to work out of the box with Feliz applications. I have been testing lots and lots of variant function components in Feliz and tracked down some cases where it doesn't work (and instead triggers full page refresh):
 - [react-refresh-webpack-plugin/issues/260](https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/260)
 - [react-refresh-webpack-plugin/issues/258](https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/258)

These two issues look different but actually, both of them can be fixed when omitting the unnecessary `export` modifier *within the same file*. It would be great if Fable was able to detect whether a value is indeed referenced outside of it's F# module/file and only add `export` when that is the case. 

### Repro code

Currently this code:
```fs
let stylesheet = Stylesheet.load "./styles/counter.css"
```
generates:
```js
import counter_2 from  "./styles/counter.css"
export const stylesheet = counter_2
```
With this feature, the generated code looks like this:
```js
import counter_2 from  "./styles/counter.css"
const stylesheet = counter_2
```
Though ideally for default imports, the intermediate value `counter_2` derived from the file name would go away but I can live with it, react-refresh seems to be OK with it and that is really specific to default imports.

This feature would also apply for (record) type definitions. Do no export unless used outside the module/file 

Right now, it is possible to decorate values with the `private` access modifier to force omitting the `export` but for record types it becomes unusable because usually you would want to use the record outside the F# module anyways. 

### Related information

* Fable version: Fable 3 - RC08
* Operating system: Windows 10

**Comment by alfonsogarciacaro** [comment]
Hmm, this is going to be tricky. The only way I think we can detect this is by scanning all the files that come after that one which, besides being computationally expensive, is not possible right now because the Fable compilation of each file is done in parallel not in sequence.

Maybe we could let the ReactComponent plugin scan the declarations of the current file and if it detects a public declaration that is not the component itself, raises a warning recommending the user to put utility functions within a private module.

**Comment by Zaid-Ajaj** [comment]
> besides being computationally expensive, is not possible right now because the Fable compilation of each file is done in parallel not in sequence.

I was afraid this might be the case but I am thinking it might be possible to find this out at the FCS level to quickly get the number of references of values across projects 🤔 the idea is to omit `export` when that number > 0 excluding the locally defined values (in the same file). 

A plugin won't do either because it has to work for exported types too 😞

**Comment by alfonsogarciacaro** [comment]
Good point! Though I'd have to check, thing is Fable uses a custom FCS build by @ncave that precisely removes symbol information, lol.

Fable AST retains much information about each declaration: if it's a type, if it's public, etc. So letting the plugin inspect the whole file and not only the declaration with the attribute should work. If exporting records is an issue we could raise a warning to recommend to use an anonymous record instead.

It may be a good idea to raise warnings and let the user do the necessary changes and get used to the patterns. If we try to do too much magic there's a risk a function is actually used somewhere else and exported, react-refresh stops working and users don't understand why.

**Comment by Zaid-Ajaj** [comment]
The problem is that the issue with exported types only happens when the types are exported *and* used from the same module. If they are used from outside the module, the issue no longer happens because react-refresh is only refreshing the module that was changed (not the module where the types where defined)

There were already enough restrictions on what React function components could take as input and how they are used. With this Fable release I wanted to remove as many of those exceptions as possible and these are the few edge cases that make huge difference IMO

I guess I will out for feedback from the issues I got on the webpack plugin repo, if it can be fixed there, it would be really huge!

**Comment by inosik** [comment]
Maybe I'm missing something, but can't you make the value `private` @Zaid-Ajaj?

**Comment by Zaid-Ajaj** [comment]
@inosik I *can* make the value `private` and that works great but with it comes to record types, I want only to export them when I actually start using them from outside the module. So basically if you are working with props defined as record and using them in the same file, you would have to set the record type private until you are working with it in another module, in which case you have to remove the `private` identifier. 

This of course would be a really annoying UX and I was hoping the compiler could do that automatically *or* they fix the issue on react-refresh where it does a full page refresh unnecessarily just because the module exports a type

**Comment by Zaid-Ajaj** [comment]
@alfonsogarciacaro After more testing an debugging, it seems that this feature is becoming more needed. Many edge cases arise in react-refresh when values and functions are unnecessarily exported and cause a full page refresh. Take [this sandbox example](https://codesandbox.io/s/react-refresh-demo-forked-9dlqi?file=/src/App.js) with the following content:
```js
import React from "react";

import "./styles.css";

const mutate = (x) => x * 10;

export default function App() {
  const [counter, setCounter] = React.useState(0);

  React.useEffect(() => {
    let interval = setInterval(() => {
      setCounter((counter) => counter + 1);
    }, 1000);

    return () => {
      clearInterval(interval);
    };
  });

  return (
    <div className="App">
      <h1>Counting at: {mutate(counter)}</h1>
      <h2>It doesn't lose state on changing things!</h2>
    </div>
  );
}
```
while the component `App` is changing its internal state, we can modify the `mutate` function to something like `(x) => x * 20` and react-refresh will successfully apply a partial re-render, but once the `mutate` function is marked as `export const` then the re-render becomes a full page re-render and we lose the state. Unless, again we make the `mutate` function upper-case: `Mutate` and then it gets picked up. 

Implementing this feature would make the react-refresh UX a lot better instead of now having to mark everything as private, move records out to another file.

**Comment by alfonsogarciacaro** [comment]
With the new plugin that gives you a handle to the file you should be able to scan for public (thus exported) declarations. It's not possible to modify them, but in any case I would strongly recommend to just raise a warning and ask the user to put helpers in a private module instead of doing too much compiler magic.

**Comment by Zaid-Ajaj** [comment]
> instead of doing too much compiler magic

You are right, I should just document it somewhere to use more private values instead.
